File: app/src/main/java/danb/speedrunbrowser/api/objects/Genre.java
package danb.speedrunbrowser.api.objects;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.io.Serializable;
import java.lang.reflect.Type;

public class Genre implements Serializable {
    public String id;
    public String name;

    public int count;

    public static class JsonConverter implements JsonDeserializer<Genre> {

        public Genre deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            Genre g = new Genre();

            if(json.isJsonPrimitive()) {
                g.id = json.getAsString();
            }
            else {

                JsonObject obj = json.getAsJsonObject();

                g.id = obj.get("id").getAsString();
                g.name = obj.get("name").getAsString();

                if(obj.has("count"))
                    g.count = obj.get("count").getAsNumber().intValue();
            }

            return g;
        }
    }
}

File: app/src/main/java/danb/speedrunbrowser/api/objects/Genre.kt
package danb.speedrunbrowser.api.objects

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException

import java.io.Serializable
import java.lang.reflect.Type

data class Genre(
    val id: String,
    val name: String = "",

    val game_count: Int = 0
) : Serializable {

    class JsonConverter : JsonDeserializer<Genre> {

        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Genre {
            if (json.isJsonPrimitive) {
                return Genre(json.asString)
            } else {

                val obj = json.asJsonObject

                return Genre(
                    id = obj.get("id").asString,
                    name = obj.get("name").asString,

                    game_count = if(obj.has("game_count"))
                        obj.get("game_count").asNumber.toInt()
                    else
                        0
                )
            }
        }
    }

    companion object {
        val ALL_GENRES_GENRE = Genre(id = "ALL_GENRES", name = "All Genres")
    }
}
